### DISCUSSION TITLE
tutorials/voyage-part2/05-configuring-the-router

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#71] [course feedback]
# tutorials/voyage-part2/05-configuring-the-router

https://www.apollographql.com/tutorials/voyage-part2/05-configuring-the-router

**Comment by mshah100** [comment]
Hi all,
I am working on the course Voyage II: Federating the Monolith

I have progressed along tot he point where we need to configure the router.

I have configured for config.yaml to be as follows:

headers:
all:
request:
- propagate:
named: “Authorization”
include_subgraph_errors:
all: true # Propagate errors from all subgraphs

when trying to start up the router using the command below:
APOLLO_KEY=<APOLLO_KEY> APOLLO_GRAPH_REF=<APOLLO_GRAPH_REF> ./router --config config.yaml

I am getting the following error, please let me know how to address this. Up to this point my project works seamlessly. I went through the Authentication and Authorization project as well to ensure I did not miss any thing.

Have also converted the the subgraph to a supergraph:

./rover subgraph publish airlock-dwih0w@current \
  --schema ./monolith/schema.graphql \
  --name monolith \
  --routing-url http://localhost:4001 \
  --convert


I have redone all of the tutorial Federating the Monolith - Voyager II and totally not sure what is wrong. Any inputs will be much appreciated as I seem to have hit a dead end.

./router --config config.yaml
Apollo Router v1.38.0 // (c) Apollo Graph, Inc. // Licensed as ELv2 (Apollo FAQ on Licensing under Elastic License v2 (ELv2) - Apollo GraphQL Docs)

:warning: The Apollo Router requires a composed supergraph schema at startup. :warning:

:point_right: DO ONE:

Pass a local schema file with the ‘–supergraph’ option:

$ ./router --supergraph <file_path>

Fetch a registered schema from GraphOS by setting
these environment variables:

$ APOLLO_KEY=“…” APOLLO_GRAPH_REF=“…” ./router

For details, see the Apollo docs:
Setting up managed federation - Apollo GraphQL Docs

:microscope: TESTING THINGS OUT?

Download an example supergraph schema with Apollo-hosted subgraphs:
$ curl -L https://supergraph.demo.starstuff.dev/ > starstuff.graphql
Run the Apollo Router in development mode with the supergraph schema:
$ ./router --dev --supergraph starstuff.graphql

**Reply by mshah100** [reply]
Here is the githuib repo https://github.com/mshah100/odyssey-voyage-II-server
I am stuck please help provide and provide inputs.

**Reply by lizhennessy** [reply]
Hi there @mshah100 ! So sorry for the delay - I am taking a look right now. I should have an update for you very soon.

**Reply by lizhennessy** [reply]
Hi @mshah100 - this error looks to be occurring because the instance of the router is _unable to find your graph_. This could be for a few reasons - 

1. There might be a mismatch between your APOLLO_KEY value and the APOLLO_GRAPH_REF.
2. One or more might be misspelled
3. Your graph might not be published correctly in Studio.

What we can do for either case is go to Studio, and select your graph of `airlock-dwih0w@current`, and navigate to Settings. From there, we'll select "This Graph" and "API Keys."
<img width="1915" alt="image" src="https://github.com/apollographql-education/content-feedback/assets/95302380/fc23ec91-0415-492b-b8be-a85653ef0826">

I would recommend generating a new API key, and then attempting to run the router command again so it can connect to your graph. We need to be careful to paste in our APOLLO_KEY and APOLLO_GRAPH_REF with a space between the two values to make sure that they're interpreted accurately by the router.

```
APOLLO_KEY=the-value-of-the-new-api-key APOLLO_GRAPH_REF=airlock-dwih0w@current ./router --config ./config.yaml
```

Would you try that and let me know if the outcome changes at all?

Thanks!

Liz

**Reply by mshah100** [reply]
worked like a charm.  Thanks

**Reply by lizhennessy** [reply]
So glad to hear it @mshah100!!!

**Comment by elperez** [comment]
Hello Apollo team,

I am using an enterprise trial account.

I'm trying to run the router with command:
APOLLO_KEY=<apollo_key> APOLLO_GRAPH_REF=airlock-5nv5t1@current ./router --config config.yaml

I am getting bellow message which is not too informative and shows "Segmentation fault: 11"

2024-02-16T05:40:42.808677Z INFO  Apollo Router v1.40.0 // (c) Apollo Graph, Inc. // Licensed as ELv2 (https://go.apollo.dev/elv2)
2024-02-16T05:40:42.809745Z INFO  Anonymous usage data is gathered to inform Apollo product development.  See https://go.apollo.dev/o/privacy for details.
Segmentation fault: 11

Looking forward to getting a bit of assistance.

Thanks,
Enrique

**Reply by lizhennessy** [reply]
Hi there @elperez! I've never seen this error before, so I wonder if it's related to the version of router. Would you try deleting the router binary from your folder and redownloading an earlier version? Here's a command that points to version 1.37.0:

```
curl -sSL https://router.apollo.dev/download/nix/v1.37.0 | sh 
```

If that doesn't work, I'd be happy to troubleshoot further!

**Reply by elperez** [reply]
That worked. I'm now able to run the router. Thank you! :)

**Comment by neoVoda** [comment]
I keep on getting this response when hitting the query, even though i have followed the instructions correctly: {
  "data": null,
  "errors": [
    {
      "message": "HTTP fetch failed from 'monolith': 500: Internal Server Error",
      "extensions": {
        "code": "SUBREQUEST_HTTP_ERROR",
        "service": "monolith",
        "reason": "500: Internal Server Error",
        "http": {
          "status": 500
        }
      }
    },
    {
      "message": "*** you must be logged in ***",
      "extensions": {
        "code": "UNAUTHENTICATED",
        "stacktrace": [
          "GraphQLError: *** you must be logged in ***",
          "    at AuthenticationError (C:\\Users\\mashegon01\\Documents\\odyssey-voyage-II-server\\monolith\\utils\\errors.js:5:10)",
          "    at C:\\Users\\mashegon01\\Documents\\odyssey-voyage-II-server\\monolith\\index.js:40:21",
          "    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)",
          "    at async context (C:\\Users\\mashegon01\\Documents\\odyssey-voyage-II-server\\monolith\\index.js:37:28)",
          "    at async ApolloServer.executeHTTPGraphQLRequest (C:\\Users\\mashegon01\\Documents\\odyssey-voyage-II-server\\monolith\\node_modules\\@apollo\\server\\dist\\cjs\\ApolloServer.js:530:32)"
        ]
      }
    }
  ]
}

**Reply by neoVoda** [reply]
error.message = Request failed with status code 400

**Reply by mabuyo** [reply]
Hey there @neoVoda - did you add the `Authorization` header when sending the query?

>  In the bottom panel of the Explorer, open the `Headers` tab. Click `New header` and set the header key as `Authorization`. Set the value as `Bearer user-1`.

Let me know if that works. I see you ran into similar issues in the future lessons, so let's get this one sorted out first before moving on.

**Comment by kendallshearman** [comment]
Hello Apollonauts! 

Update:
curl -sSL https://router.apollo.dev/download/nix/v1.37.0 | sh
With
curl -sSL https://router.apollo.dev/download/nix/latest | sh

Avoids:

ERROR: "Invalid supergraph: uses unsupported join spec version v0.4 (supported versions: v0.3, v0.2, v0.1)"

**Reply by mabuyo** [reply]
Hey @kendallshearman 👋🏽 At which step do you get this error? We wanted to specify the router version to avoid any version incompatibility issues with the course at the time of writing (since using the latest version might not be guaranteed to work if the course is outdated!).

**Reply by kendallshearman** [reply]
Hi @mabuyo!  

At this stage in the tutorial:
https://www.apollographql.com/tutorials/voyage-part2/05-configuring-the-router#:~:text=Running%20the%20router%20with%20config

I received the error when trying to run: 
APOLLO_KEY=<APOLLO_KEY> APOLLO_GRAPH_REF=<APOLLO_GRAPH_REF> ./router --config config.yaml

**Reply by mabuyo** [reply]
Fixed now, thank you for letting us know!

**Reply by kendallshearman** [reply]
Sweet, thanks @mabuyo!

**Comment by raphaelspies** [comment]
Hello, I'm running into some issues when attempting to run router locally. I'm running the following command: 

`APOLLO_KEY=<APOLLO_KEY> APOLLO_GRAPH_REF=<APOLLO_GRAPH_REF> ./router --config router-config.yaml`

and getting the following error:
`ERROR  could not create the HTTP server: Address already in use (os error 48)`

I've tried the following:
- checking to see if anything is running on port 4000 (nothing is)
-Verifying the install version of router (v1.46.0, the version listed in the tutorial above)
- Verifying the installed version of rover (0.26.3, the latest as of now)
- changing the port to run the router on by adding `supergraph:
  listen: "0.0.0.0:4002"` to my router-config

Would appreciate any suggestions as to how to rectify this issue.
Thanks!
